quantum/audio/audio_avr_pwm.c

```diff
@@ -0,0 +1,248 @@
+/* Copyright 2016 Jack Humbert
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdio.h>
+#include <string.h>
+//#include <math.h>
+#if defined(__AVR__)
+  #include <avr/pgmspace.h>
+  #include <avr/interrupt.h>
+  #include <avr/io.h>
+#endif
+#include "print.h"
+#include "audio.h"
+
+#include "wait.h"
+
+//TODO: move into audio-common state
+extern bool playing_notes;
+extern float note_timbre;
+extern uint8_t voices;
+
+#define CPU_PRESCALER 8
+
+// -----------------------------------------------------------------------------
+// Timer Abstractions
+// -----------------------------------------------------------------------------
+
+//Currently we support timers 1 and 3 used at the sime time, channels A-C,
+//pins PB5, PB6, PB7, PC4, PC5, and PC6
+#if defined(C6_AUDIO)
+    #define CPIN_AUDIO
+    #define CPIN_SET_DIRECTION DDRC |= _BV(PORTC6);
```

Comment:
You could probably simplify this by only abstracting things like the `TIMSK`, `ICR`, `OCR` registers, eg. `#define AUDIO_TIMSKx TIMSK3` etc. Cf. backlight setup:

https://github.com/qmk/qmk_firmware/blob/7f23ad72c5c736b58bab16995e7b0a599268b56f/quantum/quantum.c#L1025-L1035

Reply:
thanks for the suggestion :-)
but that's not quite the intention of the original code (that i just moved here) - which is to have (optionally) two pins/speakers in parallel = two "voices"
that's also the reason for some nested ifdefs: ifdef BPIN_AUDIO inside a ifdef CPIN_AUDIO block

Reply:
Ah, right, I see now.

Can you have polyphony using only B or C pins, or do you have to have one of each? If it's the latter I still think these defines can be improved in a similar way. It seems so, as the B pins all share the same timer, and the same for the C pins.

Reply:
it's a "pick one of the three Cs, and optionally one of the three Bs too"
the 'TIMER_x_PERIOD define is the only common one, which could be moved out - but this way everything is in one place - i lean to leaving it :-)

Reply:
In that case, shouldn't there be some `#elif`s here grouping the B and C defines together (ie. making sure you can only have one B pin defined)? Then, the registers for the two audio channels can be, say, `AUDIO1_TIMSKx` and `AUDIO2_TIMSKx`, and the actual manipulation of the registers only has to be written once.

Reply:
oh, good find - the missing elif is actually a bug... i'll refactor some more :-)

Reply:
hm, it's a bit more complicated: either C as voice 1 and optional B as voice 2; or just B as voice 1...

Reply:
Here's what I mean:

```c
#if defined(C4_AUDIO) || defined(C5_AUDIO) || defined(C6_AUDIO)
    #define CPIN_AUDIO
    #define AUDIO2_TIMSK3 TIMSK3
    #define AUDIO2_DDRC DDRC
    #define AUDIO2_TCCR3A TCCR3A
    #define AUDIO2_ICR3 ICR3

    #if defined(C6_AUDIO)
        #define AUDIO2_COM3x0 COM3A0
        #define AUDIO2_COM3x1 COM3A1
        #define AUDIO2_OCIE3x OCIE3A
        #define AUDIO2_OCR3x OCR3A
        #define AUDIO2_PORTCx PORTC6
        #define AUDIO2_TIMER3_COMPx_vect TIMER3_COMPA_vect
    #elif defined(C5_AUDIO)
        #define AUDIO2_COM3x0 COM3B0
        #define AUDIO2_COM3x1 COM3B1
        #define AUDIO2_OCIE3x OCIE3B
        #define AUDIO2_OCR3x OCR3B
        #define AUDIO2_PORTCx PORTC5
        #define AUDIO2_TIMER3_COMPx_vect TIMER3_COMPB_vect
    #elif defined(C4_AUDIO)
        #define AUDIO2_COM3x0 COM3C0
        #define AUDIO2_COM3x1 COM3C1
        #define AUDIO2_OCIE3x OCIE3C
        #define AUDIO2_OCR3x OCR3C
        #define AUDIO2_PORTCx PORTC4
        #define AUDIO2_TIMER3_COMPx_vect TIMER3_COMPC_vect
    #endif
#endif

#if defined(B5_AUDIO) || defined(B6_AUDIO) || defined(B7_AUDIO)
    #define BPIN_AUDIO
    #define AUDIO1_TIMSK1 TIMSK1
    #define AUDIO1_DDRB DDRB
    #define AUDIO1_TCCR1A TCCR1A
    #define AUDIO1_ICR1 ICR1

    #if defined(B5_AUDIO)
        #define AUDIO1_COM1x0 COM1A0
        #define AUDIO1_COM1x1 COM1A1
        #define AUDIO1_OCIE1x OCIE1A
        #define AUDIO1_OCR1x OCR1A
        #define AUDIO1_PORTBx PORTB5
        #define AUDIO1_TIMER1_COMPx_vect TIMER1_COMPA_vect
    #elif defined(B6_AUDIO)
        #define AUDIO1_COM1x0 COM1B0
        #define AUDIO1_COM1x1 COM1B1
        #define AUDIO1_OCIE1x OCIE1B
        #define AUDIO1_OCR1x OCR1B
        #define AUDIO1_PORTBx PORTB6
        #define AUDIO1_TIMER1_COMPx_vect TIMER1_COMPB_vect
    #elif defined(B7_AUDIO)
        #define AUDIO1_COM1x0 COM1C0
        #define AUDIO1_COM1x1 COM1C1
        #define AUDIO1_OCIE1x OCIE1C
        #define AUDIO1_OCR1x OCR1C
        #define AUDIO1_PORTBx PORTB7
        #define AUDIO1_TIMER1_COMPx_vect TIMER1_COMPC_vect
    #endif
#endif
```

Now all these defines are simply aliases pointing to the correct registers and constants to use when a certain pin is selected, and the actual code is written only twice - once for each voice, I assume - instead of for every supported pin.

Reply:
hm, you where quicker - just pushed a commit similar to your last comment :-s
i'll rework it a little more, the nesting in your solution is nicer :-)

Reply:
ok - what are your thoughts on [5498915](https://github.com/qmk/qmk_firmware/commit/5498915cd54093b739051cb4cea9efc37729cdc3) ?

Reply:
Looking better 👍 there's still a lot of deduplication that can be done IMO. The purpose of the register operations can be documented in comments instead of the define names if need be.

Reply:
i would argue to keep the descriptive defines - since they make code that follows more readable...
otherwise it would mean a total rewrite = replacing the long-ish defines with the methods they supply, and keeping only the actual registers as defines - like your suggested code-snippet :-S

Reply:
ok - bit the (minor) bullet and refactored the whole audio_avr_pwm code according to your initial suggestion: having only defines for registers, and not whole functions
flip-side: now the code is a bit more like audio_arm_pwm :-)

best compare the final file, and not the diff -- your thoughts?

Reply:
Note to self - trying to enforce defines with pragmas like error or warning breaks builds... a lot :P